### PR TITLE
feat(state)!: statefile implementation

### PR DESCRIPTION
**Pull request by joshua-kim**
BREAKING CHANGE: data is now persisted in the apm's statefile
Refs #19

### Overview
The APM tracks metadata (what vms are installed, what was last synced, what repos to track, etc) in locally persisted storage. The use of a database is a bit painful and makes it hard to alter our data schema, and requires users to have to keep wiping their databases. This PR is part of an ongoing series to entirely eliminate the use of leveldb in favor of tracking minimal amounts of information in the apm's `statefile`. Ideally, the apm would only track its tracked sources (what repositories it's in sync with and what version was last sync'd), and what vm aliases are installed in its statefile.

### Example

The APM's statefile now only contains:
1. `sources` - the tracked repos, and what commit they're currently synced to
2. `installation-registry` - the vms that apm knows are installed and what commit version they're using

Instead of serializing vm/subnet definitions into a database, we just use the yaml files by directly looking them up on disk.
```
➜  apm git:(statefile) cat ~/.apm/apm.state
sources:
    ava-labs/avalanche-plugins-core:
        url: https://github.com/ava-labs/avalanche-plugins-core.git
        commit: d1301636cb1562530495106dd01a2ea150c90609
        branch: refs/heads/master
installation-registry:
    ava-labs/avalanche-plugins-core:spacesvm:
        id: sqja3uK17MJxfC7AN8nGadBw9JK5BcrsNwNynsqP5Gih8M5Bm
        commit: 81cbd264cfa7a656f0432c7d2902c55871d8af33
```

### CODE REVIEW

File: storage/repository_factory.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ type RepositoryFactory interface {
 	GetRepository(alias []byte) Repository
 }
 
+// TODO replace with fs-based implementation
```
Comment (dasconnor) [inline storage/repository_factory.go:17]: Do you have a ticket for this?

File: storage/state.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+}
+
+func NewStateFile(path string) (StateFile, error) {
+	result := NewEmptyStateFile(path)
+
+	b, err := os.ReadFile(result.path)
+	var pathError *os.PathError
+	if errors.As(err, &pathError) {
+		// need to initialize the StateFile on Commit
```
Comment (dasconnor) [inline storage/state.go:34]: do you need to do something here?

**Review by dasconnor — APPROVED** [review]
Couple small questions, but you can merge when ready

File: storage/state.go (hunk 2 above)
Comment (joshua-kim) [reply storage/state.go:34]: Yeah this needs to return an err, forgot to mark a TODO. I'll update this before I merge

**Pull request merged**